pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1018,14 +1021,17 @@ private long getHighestSequenceId(OpSendMsg op) {
     }
 
     private void releaseSemaphoreForSendOp(OpSendMsg op) {
-        if (semaphore.isPresent()) {
-            semaphore.get().release(isBatchMessagingEnabled() ? op.numMessagesInBatch : 1);
-        }
+        semaphore.ifPresent(semaphore -> semaphore.release(
+                Math.min(
+                        maxPermits - semaphore.availablePermits(),
```

Comment:
I don't understand how we are dealing with concurrency here.
When you capture the number of available permits another thread could have operated on the semaphore

Reply:
We are not trying to  solve all of the issues here, this is a simple overflow protection fix.
The root cause still has to be figured out but this is simpler alternative to implementing a bounded semaphore.